api/src/main/java/org/apache/iceberg/actions/RewriteManifests.java

```diff
@@ -44,6 +45,16 @@ public interface RewriteManifests
    */
   RewriteManifests rewriteIf(Predicate<ManifestFile> predicate);
 
+  /**
+   * Rewrite manifests in a given order, based on partition columns
+   *
+   * <p>If not set, manifests will be rewritten in the order of the table's spec.
+   *
+   * @param partitionSortOrder a list of partition field names
```

Comment:
Should this be partition transform names? Considering we will also have multi-arg transforms coming up I think that is probably more accurate.

Reply:
Ack - didn't know about multi-arg transforms! 

I originally had at as expecting the transform name (`sometime_day` transformed name instead of the underlying `sometime` value). Makes the implementation easier but makes the user have to know what the internally transformed names are referred to as. With multi-arg I think this no longer viable for sure. Could always accept both and check against either the value or transform name. Think that is just too convoluted?

Reply:
Hmm, I think that's probably safe to do but there will be some edge cases that may be weird. Like i have a column named "identity" :/ 

If we can guard against any ambiguous situations I think that could be fine

Reply:
I see the question in the description about 

> I would next like to have a more capable functional interface sort(row -> {... return groupingString}), but was struggling to express a Java Functionlike input that could be used on a DataSet<Row>'s data_file struct - any pointers on how to parse a Row's struct into a native Pojo and supply a function that can be used like in a UDF here would be apprecaited!

I think it aligns with my main comment, which is that we discussed the potential to expose it as a generic function but seems like this PR directly skipped it. Seems like there are some pieces missing, let me see if I can fill it up.

At API level, we could express it as a function using `PartitionData` as the transformation source:

```
default RewriteManifests sort(List<String> partitionFieldNames) {
  return sort(...) // see partitionSpec.partitionToPath for an example of how to convert the partition fields to string
}

RewriteManifests sort(Function<PartitionData, String> partitionFieldsSortStrategy) {
  throw new UnsupportedOperationException();
}

```

and when implementing in Spark, the biggest question is now how do we convert the Row `data_file.partition` to `PartititionData` so we can apply this transform and create the new string column in the data frame that is used for sorting.

I think this is achievable by creating a UDF that transforms the `data_file.partition` Row to `SparkStructLike` using `SparkStructLike.wrap(row)`, and then create a new `PartitionData` based on that, and finally apply the input function to create its string value. You can see an example of that in `PartitionsTable.toPartitionData`. 

@RussellSpitzer @nastra let me know if you agree with this general direction, or if it is becoming too convoluted.

Reply:
Meta question - since I was struggling to get that part to work, and to keep things more concise,I had intended to submit that additional interface as a separate PR, would it be better to include here in one go? 

Thank you for the pointer there! I'll give this a shot. Converting to something useable in a UDF was what was throwing me for a loop

Reply:
My feeling is that if we merge this and you end up having to refactor it a lot, then it is better for everyone to help you make the right changes here. 

But I am all for incremental changes, so if we can agree that the function-based API proposed in the original issue can be added through refactoring later, then I am totally good with that! Let's see what others think.

Reply:
Sounds good, open to anything. I don't believe there would be much of a refactor at all FWIW. Just a new api option and single branch.

Reply:
Clarifying on the signature you're suggesting

```
RewriteManifests sort(Function<PartitionData, String> partitionFieldsSortStrategy) {
  throw new UnsupportedOperationException();
}
```

I would propose to have the method operate over a `DataFile`, so that all attributes are available for scrutiny if the user wants, not just the partition information itself. 

```
RewriteManifests sort(Function<DataFile, String> partitionFieldsSortStrategy) {
  throw new UnsupportedOperationException();
}
```

So the specific implementation would be a `SparkDataFile` I think, and need to do the same kind of `toDataFile` converter?

Reply:
ah that's actually a great idea! In that case we can actually use the `SparkDataFile` to convert the `data_file` row. You can see an example of that in test like https://github.com/apache/iceberg/blob/main/spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataFile.java#L180-L188